reactor-core/src/main/java/reactor/core/publisher/SinkManyBestEffort.java

```diff
@@ -0,0 +1,385 @@
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.stream.Stream;
+
+import reactor.core.CoreSubscriber;
+import reactor.core.Exceptions;
+import reactor.core.Scannable;
+import reactor.core.publisher.Sinks.Emission;
+import reactor.util.annotation.Nullable;
+import reactor.util.context.Context;
+
+/**
+ * @author Simon Baslé
+ */
+final class SinkManyBestEffort<T> extends Flux<T> implements Sinks.Many<T>, ContextHolder,
+                                                             Scannable {
+
+	private static final Inner[] EMPTY = new Inner[0];
+	private static final Inner[] TERMINATED = new Inner[0];
+
+	static final <T> SinkManyBestEffort<T> createBestEffort() {
+		return new SinkManyBestEffort<>(false);
+	}
+
+	static final <T> SinkManyBestEffort<T> createAllOrNothing() {
+		return new SinkManyBestEffort<>(true);
+	}
+
+	final boolean allOrNothing;
+
+	/**
+	 * Stores the error that terminated this sink, for immediate replay to late subscribers
+	 */
+	Throwable error;
+
+	volatile     Inner<T>[]                                               subscribers;
+	@SuppressWarnings("rawtypes")
+	static final AtomicReferenceFieldUpdater<SinkManyBestEffort, Inner[]> SUBSCRIBERS =
+			AtomicReferenceFieldUpdater.newUpdater(SinkManyBestEffort.class, Inner[].class, "subscribers");
+
+	SinkManyBestEffort(boolean allOrNothing) {
+		this.allOrNothing = allOrNothing;
+		SUBSCRIBERS.lazySet(this, EMPTY);
+	}
+
+	public Context currentContext() {
+		return Operators.multiSubscribersContext(subscribers);
+	}
+
+	@Override
+	public Stream<? extends Scannable> inners() {
+		return Stream.of(subscribers);
+	}
+
+	@Nullable
+	@Override
+	public Object scanUnsafe(Attr key) {
+		if (key == Attr.TERMINATED) return subscribers == TERMINATED;
+		if (key == Attr.ERROR) return error;
+
+		return null;
+	}
+
+	@Override
+	public Emission tryEmitNext(T t) {
+		Objects.requireNonNull(t, "tryEmitNext(null) is forbidden");
+
+		Inner<T>[] subs = subscribers;
+
+		if (subs == EMPTY) {
+			return Emission.FAIL_ZERO_SUBSCRIBER;
+		}
+		if (subs == TERMINATED) {
+			return Emission.FAIL_TERMINATED;
+		}
+
+		if (allOrNothing) {
+			long commonRequest = Long.MAX_VALUE;
+			for (Inner<T> sub : subs) {
+				long subRequest = sub.requested;
+				if (sub.isCancelled()) {
+					continue;
+				}
+				if (subRequest < commonRequest) {
+					commonRequest = subRequest;
+				}
+			}
+			if (commonRequest == 0) {
+				return Emission.FAIL_OVERFLOW;
+			}
+		}
+
+		int expectedEmitted = subs.length;
+		int emittedCount = 0;
+		int cancelledCount = 0;
+		for (Inner<T> sub : subs) {
+			if (sub.isCancelled()) {
+				cancelledCount++;
+				continue;
+			}
+			if (sub.tryEmitNext(t)) {
+				emittedCount++;
+			}
+			else if (sub.isCancelled()) {
+				cancelledCount++;
+			}
+		}
+		if (cancelledCount == expectedEmitted) {
+			return Emission.FAIL_ZERO_SUBSCRIBER; //TODO should it be FAIL_CANCELLED ? but technically the sink would still be usable
```

Comment:
should we even check for cancelled? I assume it would get removed, and we will only get a cancelled instance due to a race, but that's okay to still push things to a cancelled subscription since `cancel()` is not guaranteed to be instantly delivered

Reply:
it is easy enough to check for it and I think it makes sense to unnecessarily avoid pushing elements when we know the downstream is cancelled

Reply:
updated